### ISSUE TITLE
./boringproxy-linux-arm failed to create client

### ISSUE THREAD

**Issue by Zygodon** [boringproxy/boringproxy#55]
Hi, I seem to have a talent for finding problems. I have a tunnel which appears to be correctly specified in the web interface/tunnels:

camera.thewensleyhouse.xyz:41389 -> thecamera -> 127.0.0.1:5000

But I cannot start the client on the remote machine:

pi@Cedd:~ $ ./boringproxy-linux-arm client -server camera.thewensleyhouse.xyz -user admin -token zBTuQ9bjV9lfVtjZVSVIamiiIlkQnWUw -client-name thecamera
Failed to create client. Ensure the server is running. URL: https://camera.thewensleyhouse.xyz/api/users/admin/clients/thecamera

The server is in fact running:
pi@raspberrypi:~ $ ./boringproxy_arm server
2021/02/10 16:32:37 Starting up
2021/02/10 16:32:38 Enter Admin Domain: 
camera.thewensleyhouse.xyz
2021/02/10 16:33:04 Successfully acquired admin certificate
2021-02-10T16:33:19Z 127.0.0.1 GET camera.thewensleyhouse.xyz /
2021-02-10T16:33:20Z 127.0.0.1 GET camera.thewensleyhouse.xyz /logo.png
2021-02-10T16:38:16Z 127.0.0.1 GET camera.thewensleyhouse.xyz /login/
2021-02-10T16:38:16Z 127.0.0.1 GET camera.thewensleyhouse.xyz /logo.png
2021-02-10T16:38:20Z 127.0.0.1 GET camera.thewensleyhouse.xyz /
2021-02-10T16:38:20Z 127.0.0.1 GET camera.thewensleyhouse.xyz /logo.png
2021-02-10T16:41:58Z 127.0.0.1 GET camera.thewensleyhouse.xyz /logo.png
2021-02-10T16:43:03Z 127.0.0.1 GET camera.thewensleyhouse.xyz /logo.png

(The reason there are so many GETs is because I had difficulty logging in to the web interface unless I use /login?access_token=... with the address - I can't get it to prompt me for a token, so I needed several attempts before getting it right - but that's a different issue)

**Comment by anderspitman** [comment]
> Hi, I seem to have a talent for finding problems

That's a useful talent!

I'll think about this and get back to you. In the mean time I just wanted to mention a couple things:

* You probably don't want to post your actual server address publicly here, and definitely no any real tokens. Not sure if they ones you shared are real or not but just making sure.
* For the login issues, I agree it's a bit clunky. I ran into some issues with redirects early on and just gave up, but I need to revisit it. For now, what you probably want is to visit `/#/tunnels` on the admin domain.

**Comment by Zygodon** [comment]
Thanks, @anderspitman. Yes, that was a bit of a slip (posting something like an actual address/token). My intention once I've got everything working and understood, is to wipe the machines and start again from scratch for deployment - but I take your point.

I'll be investigating my problem further today and will let you know if I find anything significant.

**Comment by Zygodon** [comment]
Hi @anderspitman, I can't get a handle on this problem at all. I deleted the executables on server and client, and the .json db file on the executable, then downloaded . -linux-arm version again onto both machines so I'm sure they have the same version. I've checked the setcap on the server. The server works fine and I can access the web interface. But I cannot get the client to run.

This is particularly annoying because clearly at one time it was, in fact, running, otherwise I would not have the tunnel, as shown in my original post (of course, I've lost that now that I have reinstated the executables). The only way I would have made that tunnel was from the client as described in your video. I'll go away and smoke something while I think about it ...

**Comment by anderspitman** [comment]
@Zygodon I think I might know the problem. Your admin-domain and tunnel domain need to be different. ie when you run the client, `-server` needs to be a different value from the tunnel domain.

The reason is because the server uses the domains to know whether to handle the request itself (such as for the UI) or proxy it to a tunnel.

**Comment by anderspitman** [comment]
There's obviously room for improved documentation here.

**Comment by Zygodon** [comment]
The sub-domains are different. Like in your demo (form memory) your apex (?) domain is brng.pro; the server is bpdemo.brng.prp and the client is etherpad.brng.pro. In my case I have the apex domain as shown above with subdomain camera for the server; then I was using thecamera as the client name (that's too confusing, I use "my_client" now but the pronciple is the same).

Is that not OK?

**Comment by Zygodon** [comment]
Apologies for the typos.

**Comment by anderspitman** [comment]
When you start the client you're using `-server camera.thewensleyhouse.xyz`. I think you need `-server thewensleyhouse.xyz`, or whatever your admin-domain is. They tunnel domain and -server can't be the same.

**Comment by Zygodon** [comment]
I must be misunderstanding something. I deleted boringproxy_db.json to persuade boringproxy that it was starting afresh. I started boringproxy server and gave it domain name camera.thewensleyhouse.xyz. I was able to open the web interface at https://camera.thewensleyhouse.xyz. with the new token of course.

Over to the client machine, ./boringproxy client -server thewensleyhouse.xyz -token the new token etc.

Still get "Failed to create client".

**Comment by anderspitman** [comment]
What is the value of `admin-domain` when you start the server?

**Comment by Zygodon** [comment]
camera.thewensleyhouse.xyz

(I've tried using just thewensleyhouse.xyz on first starting the server, but then the system knows nothing about a camera sub domain and I don't know where to give it that information. And I still can't start the client!)

**Comment by Zygodon** [comment]
Both server and client are inside my home router (NAT). I've port forwarded the server's port 80 to the router, and in Namecheap I've got the records pointed at the public address of the router. But - I guess, WITHIN my local network, the potential client doesn't know about that ...?

**Comment by anderspitman** [comment]
> camera.thewensleyhouse.xyz
> 
> (I've tried using just thewensleyhouse.xyz on first starting the server, but then the system knows nothing about a camera sub domain and I don't know where to give it that information. And I still can't start the client!)

Yeah I'm pretty sure this is your problem. They need to be different. What do you mean by "the system"? Are you referring to boringproxy, or your DNS setup, or something else? The way that you tell boringproxy about the subdomain is by using `camera.thewensleyhouse.xyz` when you create the tunnel in the web UI.

**Comment by anderspitman** [comment]
> Both server and client are inside my home router (NAT). I've port forwarded the server's ports 80 and 443 to the router, and in Namecheap I've got the records pointed at the public address of the router. But - I guess, WITHIN my local network, the potential client doesn't know about that ...?

That should work fine. But since you're hosting directly from home, have you considered using a traditional reverse proxy like Caddy (or nginx etc) instead? boringproxy is specifically designed for running on a public server like a VPS, since that's when it gives you the most advantages.

**Comment by anderspitman** [comment]
ps you don't necessarily have to use `thewensleyhouse.xyz` for the admin domain. You can also use something like `admin.thewensleyhouse.xyz`. You just can't use the same domain for any of your tunnels.

**Comment by Zygodon** [comment]
Aha, that could be helpful, thank you; I'll try again tomorrow. Boringproxy is best for me, I think, because ultimately the client will be at a truly remote location, communicating on the cell phone network. The set up at the moment is just for development where I have immediate access to both machines. And anyway it's cheaper to get it running on the WiFi, I don't want to use more data on my SIM than necessary.

**Comment by anderspitman** [comment]
That makes sense. Another feature I plan added eventually is control over how often the client polls the server for tunnel updates. Currently it's quite chatty (every 2 seconds), which will needlessly use up data in your use case. I think often tunnels don't change much, and you could get away with minutes or even hours between polling, or even just checking at startup.

**Comment by Zygodon** [comment]
I'll keep in touch about that once I'm in a deployment situation. Meanwhile I'm still hacking away at my problem - only now I find server cannot get a certificate for admin.thewensleyhouse.xyz. Looks like we're back to the problem before last! I've deleted boringproxy_db.json and made a new, empty authorized_keys file in /.ssh. 
Is deleting the db.json file an adequate way of getting back to square one? Else, how can I get certificates for thewensleyhouse.xyz and camera.etc, but not admin.etc?
So I'm going to leave it for now!

Cheers, @anderspitman, and thanks for your help

**Comment by anderspitman** [comment]
You can also try deleting `$HOME/.local/share/certmagic`, but I'd be surprised if that changes anything. certmagic has been solid about keep its state in good shape.

Are the permissions on authorized_keys 600?

You're sure your DNS is setup up with a proper wildcard, ie A record `*.thewensleyhouse.xyz` -> your router IP?

**Comment by Zygodon** [comment]
Hi @anderspitman. Today I've made progress and I can get the client running. Thank you for your suggestion about having an admin subdomain, that was a great help in clarifying my understanding and I think was the crucial step in solving the immediate problem. For anybody else who may be experiencing similar difficulties, I found your advice to delete $HOME/.local/share/certmagic useful in getting me completely reset to a clean start, as well as creating a new, empty authorized_tokens file with the appropriate permissions (600).

So, I consider this issue resolved, if you would like to close it. Unfortunately, though, I've still not quite jumped through all the hoops; I'll start a new issue to deal with my next problem.

As ever, thanks for your help and patience.
